### ISSUE TITLE
Double margins in some layouts

### ISSUE THREAD

**Issue by jx11r** [qtile/qtile#3337]
**Layouts with this issue:**
>  - Columns
>  - ~Spiral~ (Fixed: #3331)
>  - Stack
>  - Bsp
>  - Matrix
>  - RatioTile
>  - Tile
>  - VerticalTile
>  - Zoomy

**Layouts without this issue:**
>  - MonadTall
>  - MonadWide
>  - MonadThreeCol

**Additional context:**
```python
config = {
    'margin': 30,
    'single_margin': 30,
}

layouts = [
    layout.MonadTall(**config),
    layout.Columns(**config),
]
```
<details>
  <summary>Screenshot</summary>
  <h3 align="center">MonadTall</h3>
  <img src="https://user-images.githubusercontent.com/78786202/163640333-405fb8e9-fb01-400b-a323-b9f6149abf80.png" />
  <h3 align="center">Columns</h3>
  <img src="https://user-images.githubusercontent.com/78786202/163640491-59d73a5a-0b10-4aa4-9c9d-4f56042d6c6e.png" />
</details>

<details>
  <summary>Screenshot (with grids)</summary>
  <h3 align="center">MonadTall</h3>
  <img src="https://user-images.githubusercontent.com/78786202/163648947-3820d1c4-4c35-454c-aea1-2751a7578b3c.png" />
  <h3 align="center">Columns</h3>
  <img src="https://user-images.githubusercontent.com/78786202/163648957-98aabc85-03f2-454f-8d35-d653e1112d6a.png" />
</details>

**Comment by elParaguayo** [comment]
Thanks. I noticed this too. 

I think what's happening is that we get doubling of margins when windows are next to each other in some layouts.

Will take a look.

**Comment by elParaguayo** [comment]
OK - the xmonad layout does have additional handling for this:
https://github.com/qtile/qtile/blob/c1f5073935a61f2bb9a36e198791590410ec07a1/libqtile/layout/xmonad.py#L358

Will check how this works and then see if we can add to the other layouts in a consistent way.

**Comment by jx11r** [comment]
Yup, it looks exactly like the margin is doubled on the other layouts.

Thanks for checking it out.

**Comment by arjan-s** [comment]
Similar but more specific issue here: https://github.com/qtile/qtile/issues/3193

**Comment by elParaguayo** [comment]
@arjan-s that's a different issue. This is about doubling margins when windows are placed next to each other. The `single_margin` option is about setting a different margin when there's only one window.

**Comment by arjan-s** [comment]
@elParaguayo Oops, sorry, you're right!

**Comment by elParaguayo** [comment]
@justinesmithies - I think this is the issue you mentioned on IRC yesterday.

**Comment by justinesmithies** [comment]
Yes that's the same issue thanks for the heads up.

**Comment by mbenford** [comment]
> OK - the xmonad layout does have additional handling for this:
> 
> https://github.com/qtile/qtile/blob/c1f5073935a61f2bb9a36e198791590410ec07a1/libqtile/layout/xmonad.py#L358
> 
> Will check how this works and then see if we can add to the other layouts in a consistent way.

@elParaguayo MonadThreeCol solves this in a [different way](https://github.com/qtile/qtile/blob/c1f5073935a61f2bb9a36e198791590410ec07a1/libqtile/layout/xmonad.py#L1141) . When I was writing it I tried to come up with a more general approach that relies on the window position instead of its index. I planned on applying that idea to all xmonad layouts, but I didn't have enough time to work on it back then. :(

**Comment by thingsiplay** [comment]
I am not sure if this is the same issue, but I noticed that not all layouts have the same gaps. I have setup bunch of layouts and use a default `'margin': 2`. Only the Monad* and Spiral layouts require `layout_default['margin'] + 2` to look the same. I have a 1440p resolution.

Need correction:
* MonadTall, MonadThreeCol, MonadWide, Spiral

Works with my default value:
* Bsp, Matrix, Tile, Columns, RatioTile, Stack, VerticalTile

Some notes I noticed while testing:

* On MonadThreeCol the left window gap on its right to middle window has a smaller gap than the window on the right most side to middle window. 
* And for the "corrected" layouts with +2 margin the margin between the top bar and all windows and on all sides of windows to the edge of screen also have a bit bigger margin. 
* For Matrix and Columns it is also not the same. I have currently 3 windows open. Matrix and Columns look basically identical, besides the gap between window 2 and 3 on Columns is bigger than on Matrix. But the gap between window 1 and 2 is the same. Both have the same margin=2 setting without correction.

**Comment by mbenford** [comment]
@thingsiplay I think the Monad layouts are actually the right ones when it comes to gaps because they handle double margins, whereas the others you mentioned don't, AFAIK. 

I'll look into MonadThreeCol, though. It should calculate all margins consistently.

**Comment by thingsiplay** [comment]
I think Columns have the same issue as MonadThreeCol with the third window on the right side. My above description is not very clear, struggling with writing good reports.

**Comment by jx11r** [comment]
> Works with my default value:
> 
>     * Bsp, Matrix, Tile, Columns, RatioTile, Stack, VerticalTile
> 

Actually those are the layouts that have duplicate margins, just look at the screenshots.

**Comment by thingsiplay** [comment]
You are right. At least I know this is the same issue now and my observation was right, just my conclusion was wrong. I was so much fixed on the idea that if most are acting like this, then it is right and the few outliers are the wrong.

**Comment by mbenford** [comment]
@thingsiplay I measured all margins in MonadThreeCol with GIMP and they all had exactly the same size (I used a 5-pixel margin). Could you provide more information about the issue on your end?

@elParaguayo I think double margins can be fixed for all layouts by having a base `place` method that adjusts margins before calling `client.place`. This method would take into account the window _position_ on the screen instead of its _index_:

```python
    def _place_client(self, client, left, top, width, height):
        """Place a client on the screen

        Will prevent double margins by applying east and south margins only
        when the client is the rightmost or the bottommost window.
        """
        rightmost = left + width - self.screen_rect.x >= self.screen_rect.width
        bottommost = top + height - self.screen_rect.y >= self.screen_rect.height
        margin = [self.margin] * 4
        if not rightmost:
            margin[1] = 0
        if not bottommost:
            margin[2] = 0

        client.place(
            left,
            top,
            width - 2 * self.border_width,
            height - 2 * self.border_width,
            self.border_width,
            self.border_focus if client.has_focus else self.border_normal,
            margin=margin,
        )
```

This is in use in MonadThreeCol and it seems to work just fine. Do you think it could be a general solution for all layouts?

**Comment by elParaguayo** [comment]
I'd much prefer to have a single solution than implementing something on a case by case basis.

**Comment by thingsiplay** [comment]
@mbenford You are right, it is fine. Turns out it was the uneven ratio of 0.42 on my side. Overall there seem to be no problem regarding the third column. Sorry for the confusion.

**Comment by jx11r** [comment]
Uh, I thought we'd get another 90 days of inactivity to keep this issue from getting closed.
